File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/mail_detail/components/BookletPager.kt
@file:Suppress(
    "PackageNaming",
    "MagicNumber",
    "LongMethod",
    "TooManyFunctions",
    "LongParameterList",
    "ModifierMissing",
)

package app.pantopus.android.ui.screens.mailbox.mail_detail.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import coil.compose.AsyncImage
import kotlinx.coroutines.launch

/**
 * T6.5c (P21) — Android twin of iOS `BookletPager`. Two render modes:
 *  - **page** — `HorizontalPager` over the page images with a control
 *    strip (prev / page-N-of-M / next + scrubber + "view all pages").
 *  - **grid** — 3-column thumbnail grid; tap a thumbnail to switch back
 *    to page mode at that page.
 */

enum class BookletPagerMode { Page, Grid }

const val BOOKLET_PAGER_TAG = "bookletPager"

@Composable
fun BookletPager(
    pages: List<String>,
    modifier: Modifier = Modifier,
    initialPage: Int = 0,
    initialMode: BookletPagerMode = BookletPagerMode.Page,
) {
    if (pages.isEmpty()) return
    var mode by rememberSaveable { mutableStateOf(initialMode) }
    val pagerState = rememberPagerState(
        initialPage = initialPage.coerceIn(0, pages.size - 1),
    ) { pages.size }
    val scope = rememberCoroutineScope()

    Column(
        modifier = modifier.fillMaxWidth().testTag(BOOKLET_PAGER_TAG),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        when (mode) {
            BookletPagerMode.Page ->
                PageMode(
                    pages = pages,
                    state = pagerState,
                    onPrev = {
                        if (pagerState.currentPage > 0) {
                            scope.launch { pagerState.animateScrollToPage(pagerState.currentPage - 1) }
                        }
                    },
                    onNext = {
                        if (pagerState.currentPage < pages.size - 1) {
                            scope.launch { pagerState.animateScrollToPage(pagerState.currentPage + 1) }
                        }
                    },
                    onShowGrid = { mode = BookletPagerMode.Grid },
                )
            BookletPagerMode.Grid ->
                GridMode(
                    pages = pages,
                    currentPage = pagerState.currentPage,
                    onJump = { idx ->
                        scope.launch { pagerState.scrollToPage(idx) }
                        mode = BookletPagerMode.Page
                    },
                    onBackToReader = { mode = BookletPagerMode.Page },
                )
        }
    }
}

@Composable
private fun PageMode(
    pages: List<String>,
    state: androidx.compose.foundation.pager.PagerState,
    onPrev: () -> Unit,
    onNext: () -> Unit,
    onShowGrid: () -> Unit,
) {
    HorizontalPager(
        state = state,
        modifier = Modifier.fillMaxWidth().height(360.dp),
        pageSpacing = Spacing.s3,
    ) { idx ->
        BookletPageImage(
            url = pages[idx],
            modifier = Modifier.padding(horizontal = Spacing.s4).testTag("bookletPager_page_$idx"),
        )
    }
    PageIndicator(
        currentPage = state.currentPage,
        totalPages = pages.size,
        onPrev = onPrev,
        onNext = onNext,
        onShowGrid = onShowGrid,
    )
}

@Composable
private fun PageIndicator(
    currentPage: Int,
    totalPages: Int,
    onPrev: () -> Unit,
    onNext: () -> Unit,
    onShowGrid: () -> Unit,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(
                    width = 1.dp,
                    color = PantopusColors.appBorder,
                    shape = RoundedCornerShape(Radii.lg),
                )
                .padding(Spacing.s3),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            CircularIconButton(
                icon = PantopusIcon.ChevronLeft,
                contentDescription = "Previous page",
                tag = "bookletPager_prev",
                enabled = currentPage > 0,
                onClick = onPrev,
            )
            Row(
                modifier = Modifier.weight(1f),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = "Page ${currentPage + 1}",
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appText,
                    modifier = Modifier.testTag("bookletPager_pageLabel"),
                )
                Spacer(Modifier.width(4.dp))
                Text(
                    text = "of $totalPages",
                    fontSize = 13.sp,
                    color = PantopusColors.appTextSecondary,
                )
            }
            CircularIconButton(
                icon = PantopusIcon.ChevronRight,
                contentDescription = "Next page",
                tag = "bookletPager_next",
                enabled = currentPage < totalPages - 1,
                onClick = onNext,
            )
        }
        Scrubber(currentPage = currentPage, totalPages = totalPages)
        Box(modifier = Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
            Row(
                modifier =
                    Modifier
                        .clip(RoundedCornerShape(Radii.pill))
                        .background(PantopusColors.primary50)
                        .clickable(onClick = onShowGrid)
                        .padding(horizontal = Spacing.s3, vertical = Spacing.s1)
                        .testTag("bookletPager_toggleGrid")
                        .semantics { contentDescription = "View all pages" },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.FileType,
                    contentDescription = null,
                    size = 12.dp,
                    tint = PantopusColors.primary600,
                )
                Text(
                    text = "View all pages",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.primary600,
                )
            }
        }
    }
}

@Composable
private fun CircularIconButton(
    icon: PantopusIcon,
    contentDescription: String,
    tag: String,
    enabled: Boolean,
    onClick: () -> Unit,
) {
    Box(
        modifier =
            Modifier
                .size(32.dp)
                .clip(CircleShape)
                .background(PantopusColors.appSurfaceSunken)
                .clickable(enabled = enabled, onClick = onClick)
                .testTag(tag)
                .semantics { this.contentDescription = contentDescription },
        contentAlignment = Alignment.Center,
    ) {
        PantopusIconImage(
            icon = icon,
            contentDescription = null,
            size = 14.dp,
            tint = if (enabled) PantopusColors.appTextStrong else PantopusColors.appTextMuted,
        )
    }
}

@Composable
private fun Scrubber(currentPage: Int, totalPages: Int) {
    val ratio =
        if (totalPages <= 1) 1f else currentPage.coerceAtMost(totalPages - 1).toFloat() / (totalPages - 1)
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .height(4.dp)
                .clip(RoundedCornerShape(2.dp))
                .background(PantopusColors.appSurfaceSunken),
        contentAlignment = Alignment.CenterStart,
    ) {
        Box(
            modifier =
                Modifier
                    .fillMaxWidth(ratio.coerceAtLeast(0.05f))
                    .height(4.dp)
                    .clip(RoundedCornerShape(2.dp))
                    .background(PantopusColors.primary600),
        )
    }
}

@Composable
private fun GridMode(
    pages: List<String>,
    currentPage: Int,
    onJump: (Int) -> Unit,
    onBackToReader: () -> Unit,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg)),
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(1.dp)) {
                Text(
                    text = "ALL PAGES",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Bold,
                    letterSpacing = 0.5.sp,
                    color = PantopusColors.appTextSecondary,
                )
                Text(
                    text = "Tap a thumbnail to jump there",
                    fontSize = 11.sp,
                    color = PantopusColors.appTextSecondary,
                )
            }
            Text(
                text = "${pages.size} pages",
                modifier =
                    Modifier
                        .clip(RoundedCornerShape(Radii.pill))
                        .background(PantopusColors.appSurfaceSunken)
                        .padding(horizontal = Spacing.s2, vertical = 4.dp),
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appTextStrong,
            )
        }
        LaunchedEffect(Unit) { /* layout hook for tests */ }
        LazyVerticalGrid(
            columns = GridCells.Fixed(3),
            modifier = Modifier.fillMaxWidth().height(360.dp).padding(Spacing.s3),
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            verticalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            items(pages.size) { idx ->
                ThumbnailCell(
                    url = pages[idx],
                    page = idx + 1,
                    isCurrent = idx == currentPage,
                    onClick = { onJump(idx) },
                )
            }
        }
        Box(
            modifier = Modifier.fillMaxWidth().padding(bottom = Spacing.s3),
            contentAlignment = Alignment.Center,
        ) {
            Row(
                modifier =
                    Modifier
                        .clip(RoundedCornerShape(Radii.pill))
                        .background(PantopusColors.primary50)
                        .clickable(onClick = onBackToReader)
                        .padding(horizontal = Spacing.s3, vertical = Spacing.s1)
                        .testTag("bookletPager_togglePage")
                        .semantics { contentDescription = "Back to reader" },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.ChevronLeft,
                    contentDescription = null,
                    size = 12.dp,
                    tint = PantopusColors.primary600,
                )
                Text(
                    text = "Back to reader",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.primary600,
                )
            }
        }
    }
}

@Composable
private fun BookletPageImage(url: String, modifier: Modifier = Modifier) {
    AsyncImage(
        model = url,
        contentDescription = "Booklet page",
        modifier =
            modifier
                .fillMaxWidth()
                .aspectRatio(3f / 4f)
                .clip(RoundedCornerShape(Radii.lg))
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg)),
        contentScale = ContentScale.Fit,
    )
}

@Composable
private fun ThumbnailCell(
    url: String,
    page: Int,
    isCurrent: Boolean,
    onClick: () -> Unit,
) {
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .aspectRatio(3f / 4f)
                .clip(RoundedCornerShape(6.dp))
                .border(
                    width = if (isCurrent) 2.5.dp else 1.dp,
                    color = if (isCurrent) PantopusColors.primary600 else PantopusColors.appBorder,
                    shape = RoundedCornerShape(6.dp),
                )
                .clickable(onClick = onClick)
                .testTag("bookletPager_thumb_${page - 1}")
                .semantics { contentDescription = "Jump to page $page" },
    ) {
        AsyncImage(
            model = url,
            contentDescription = null,
            modifier = Modifier.fillMaxSize().background(PantopusColors.appSurfaceSunken),
            contentScale = ContentScale.Fit,
        )
        if (isCurrent) {
            Box(
                modifier =
                    Modifier
                        .padding(4.dp)
                        .size(18.dp)
                        .clip(CircleShape)
                        .background(PantopusColors.primary600)
                        .align(Alignment.TopEnd),
                contentAlignment = Alignment.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.Eye,
                    contentDescription = null,
                    size = 10.dp,
                    tint = PantopusColors.appTextInverse,
                )
            }
        }
        Text(
            text = "$page",
            modifier =
                Modifier
                    .align(Alignment.BottomCenter)
                    .padding(bottom = 4.dp),
            fontSize = 10.sp,
            fontWeight = FontWeight.Bold,
            fontStyle = FontStyle.Italic,
            fontFamily = FontFamily.Serif,
            color = PantopusColors.appTextSecondary,
        )
    }
}

File: frontend/apps/ios/Pantopus/Features/Mailbox/MailDetail/Components/BookletPager.swift
//
//  BookletPager.swift
//  Pantopus
//
//  T6.5c (P21) — Booklet variant of the A17 shell's body slot.
//
//  Two render modes per `booklet.jsx`:
//    - `.page` — indicator strip ("Page N of M" + sky scrubber +
//      grid-mode toggle button) above a full-width `TabView` swiping
//      through folded paper page images.
//    - `.grid` — 3-column thumbnail grid; tap a thumbnail to jump
//      straight back to `.page` mode at that page.
//
//  Lives in the feature folder per the P21 brief — the swipeable
//  paper-page geometry doesn't generalise to other mail variants.
//  Community attachments (P22) use a different shape.
//

import SwiftUI

/// Rendering mode for the pager. Exposed at the view init so the variant
/// view can persist the mode in its caller (e.g. across re-renders).
public enum BookletPagerMode: Sendable, Hashable {
    case page
    case grid
}

@MainActor
public struct BookletPager: View {
    private let pages: [URL]
    private let pageCount: Int
    @State private var currentIndex: Int
    @State private var mode: BookletPagerMode

    public init(
        pages: [URL],
        pageCount: Int? = nil,
        initialPage: Int = 0,
        initialMode: BookletPagerMode = .page
    ) {
        self.pages = pages
        self.pageCount = max(pageCount ?? pages.count, pages.count)
        _currentIndex = State(initialValue: max(0, min(initialPage, max(0, pages.count - 1))))
        _mode = State(initialValue: initialMode)
    }

    public var body: some View {
        VStack(spacing: Spacing.s2) {
            switch mode {
            case .page:
                pageMode
            case .grid:
                gridMode
            }
        }
        .accessibilityIdentifier("bookletPager")
    }

    // MARK: - Page mode

    private var pageMode: some View {
        VStack(spacing: Spacing.s2) {
            pageIndicator
            TabView(selection: $currentIndex) {
                ForEach(Array(pages.enumerated()), id: \.offset) { idx, url in
                    BookletPageImage(
                        url: url,
                        pageNumber: idx + 1,
                        hasNextPage: idx < pageCount - 1
                    )
                    .tag(idx)
                    .padding(.horizontal, Spacing.s4)
                    .accessibilityIdentifier("bookletPager_page_\(idx)")
                }
            }
            .tabViewStyle(.page(indexDisplayMode: .never))
            .frame(height: 420)
        }
    }

    private var pageIndicator: some View {
        VStack(spacing: Spacing.s2) {
            HStack(alignment: .center, spacing: Spacing.s2) {
                Button {
                    currentIndex = max(0, currentIndex - 1)
                } label: {
                    Icon(.chevronLeft, size: 14, color: Theme.Color.appTextStrong)
                        .frame(width: 32, height: 32)
                        .background(Theme.Color.appSurfaceSunken)
                        .clipShape(Circle())
                }
                .buttonStyle(.plain)
                .disabled(currentIndex == 0)
                .accessibilityLabel("Previous page")
                .accessibilityIdentifier("bookletPager_prev")
                VStack(alignment: .center, spacing: 1) {
                    HStack(spacing: Spacing.s1) {
                        Text("Page \(currentIndex + 1)")
                            .font(.system(size: 13, weight: .bold))
                            .foregroundStyle(Theme.Color.appText)
                        Text("of \(pageCount)")
                            .font(.system(size: 13))
                            .foregroundStyle(Theme.Color.appTextSecondary)
                    }
                    .accessibilityLabel("Page \(currentIndex + 1) of \(pageCount)")
                    .accessibilityIdentifier("bookletPager_pageLabel")
                }
                .frame(maxWidth: .infinity)
                Button {
                    currentIndex = min(pages.count - 1, currentIndex + 1)
                } label: {
                    Icon(.chevronRight, size: 14, color: Theme.Color.appTextStrong)
                        .frame(width: 32, height: 32)
                        .background(Theme.Color.appSurfaceSunken)
                        .clipShape(Circle())
                }
                .buttonStyle(.plain)
                .disabled(currentIndex >= pages.count - 1)
                .accessibilityLabel("Next page")
                .accessibilityIdentifier("bookletPager_next")
            }
            scrubber
            Button {
                mode = .grid
            } label: {
                HStack(spacing: Spacing.s1) {
                    Icon(.fileType, size: 12, color: Theme.Color.primary600)
                    Text("View all pages")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(Theme.Color.primary600)
                }
                .padding(.horizontal, Spacing.s3)
                .padding(.vertical, Spacing.s1)
                .background(Theme.Color.primary50)
                .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("bookletPager_toggleGrid")
        }
        .padding(Spacing.s3)
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
    }

    private var scrubber: some View {
        GeometryReader { proxy in
            let totalSegments = max(1, pageCount - 1)
            let filled = pageCount <= 1
                ? proxy.size.width
                : proxy.size.width * CGFloat(currentIndex) / CGFloat(totalSegments)
            ZStack(alignment: .leading) {
                Capsule()
                    .fill(Theme.Color.appSurfaceSunken)
                    .frame(height: 4)
                Capsule()
                    .fill(Theme.Color.primary600)
                    .frame(width: max(4, filled), height: 4)
            }
        }
        .frame(height: 4)
        .accessibilityHidden(true)
    }

    // MARK: - Grid mode

    private var gridMode: some View {
        VStack(alignment: .leading, spacing: Spacing.s0) {
            HStack(alignment: .center, spacing: Spacing.s2) {
                VStack(alignment: .leading, spacing: 1) {
                    Text("ALL PAGES")
                        .font(.system(size: 11, weight: .bold))
                        .tracking(0.5)
                        .foregroundStyle(Theme.Color.appTextSecondary)
                        .accessibilityAddTraits(.isHeader)
                    Text("Tap a thumbnail to jump there")
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Color.appTextSecondary)
                }
                Spacer(minLength: Spacing.s0)
                Text("\(pageCount) pages")
                    .font(.system(size: 11, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextStrong)
                    .padding(.horizontal, Spacing.s2)
                    .padding(.vertical, Spacing.s1)
                    .background(Theme.Color.appSurfaceSunken)
                    .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
            }
            .padding(.horizontal, Spacing.s3)
            .padding(.vertical, Spacing.s2)
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
            LazyVGrid(
                columns: Array(repeating: GridItem(.flexible(), spacing: Spacing.s2), count: 3),
                spacing: Spacing.s2
            ) {
                ForEach(Array(pages.enumerated()), id: \.offset) { idx, url in
                    Button {
                        currentIndex = idx
                        mode = .page
                    } label: {
                        ThumbnailCell(
                            url: url,
                            page: idx + 1,
                            isCurrent: idx == currentIndex,
                            hasNextPage: idx < pageCount - 1
                        )
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Jump to page \(idx + 1)")
                    .accessibilityIdentifier("bookletPager_thumb_\(idx)")
                }
            }
            .padding(Spacing.s3)
            HStack {
                Spacer()
                Button {
                    mode = .page
                } label: {
                    HStack(spacing: Spacing.s1) {
                        Icon(.chevronLeft, size: 12, color: Theme.Color.primary600)
                        Text("Back to reader")
                            .font(.system(size: 11, weight: .semibold))
                            .foregroundStyle(Theme.Color.primary600)
                    }
                    .padding(.horizontal, Spacing.s3)
                    .padding(.vertical, Spacing.s1)
                    .background(Theme.Color.primary50)
                    .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("bookletPager_togglePage")
                Spacer()
            }
            .padding(.bottom, Spacing.s3)
        }
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
    }
}

private struct BookletPageImage: View {
    let url: URL
    let pageNumber: Int
    let hasNextPage: Bool

    var body: some View {
        BookletPaperPageChrome(hasNextPage: hasNextPage) {
            AsyncImage(url: url) { phase in
                switch phase {
                case let .success(image):
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                case .failure:
                    fallback(icon: .alertCircle, label: "Couldn't load page")
                case .empty:
                    fallback(icon: .fileType, label: "Loading...")
                @unknown default:
                    fallback(icon: .fileType, label: "")
                }
            }
        }
        .accessibilityLabel("Booklet page \(pageNumber)")
    }

    private func fallback(icon: PantopusIcon, label: String) -> some View {
        ZStack {
            Theme.Color.appSurfaceSunken
            VStack(spacing: Spacing.s2) {
                Icon(icon, size: 24, color: Theme.Color.appTextMuted)
                if !label.isEmpty {
                    Text(label)
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Color.appTextSecondary)
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

private struct ThumbnailCell: View {
    let url: URL
    let page: Int
    let isCurrent: Bool
    let hasNextPage: Bool

    var body: some View {
        ZStack(alignment: .topTrailing) {
            BookletPaperPageChrome(
                hasNextPage: hasNextPage,
                foldSize: Spacing.s4,
                cornerRadius: 6
            ) {
                AsyncImage(url: url) { phase in
                    switch phase {
                    case let .success(image):
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .frame(maxWidth: .infinity, maxHeight: .infinity)
                    case .failure, .empty:
                        Rectangle()
                            .fill(Theme.Color.appSurfaceSunken)
                    @unknown default:
                        Rectangle()
                            .fill(Theme.Color.appSurfaceSunken)
                    }
                }
            }
            .overlay(
                RoundedRectangle(cornerRadius: 6)
                    .stroke(
                        isCurrent ? Theme.Color.primary600 : Theme.Color.appBorder,
                        lineWidth: isCurrent ? 2.5 : 1
                    )
            )
            VStack(alignment: .trailing, spacing: Spacing.s0) {
                if isCurrent {
                    Icon(.eye, size: 10, color: Theme.Color.appTextInverse)
                        .frame(width: 18, height: 18)
                        .background(Theme.Color.primary600)
                        .clipShape(Circle())
                        .padding(Spacing.s1)
                }
                Spacer(minLength: Spacing.s0)
                Text("\(page)")
                    .font(.system(size: 10, weight: .bold, design: .serif))
                    .italic()
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .padding(.horizontal, Spacing.s1)
                    .padding(.bottom, Spacing.s1)
                    .frame(maxWidth: .infinity, alignment: .center)
            }
        }
    }
}

#Preview {
    BookletPager(pages: [
        "https://placehold.co/360x480",
        "https://placehold.co/360x480/orange/white",
        "https://placehold.co/360x480/blue/white",
        "https://placehold.co/360x480/green/white",
        "https://placehold.co/360x480/red/white",
        "https://placehold.co/360x480/purple/white"
    ].compactMap(URL.init(string:)), pageCount: 6)
        .padding()
        .background(Theme.Color.appBg)
}
